### PR TITLE
Add Contact training provider to ApplicationReviewComponent

### PR DESCRIPTION
**Pull request by inulty-dfe**
## Context

As part of the update to the candidate interfaces application choices index and show, we want to give the candidate access to the contact information for the provider of the course they are applying to. We are showing the candidate the email address and the phone number of the provider on the application show page and on their offer page.

## Changes proposed in this pull request

As part of the update to the candidate interfaces application choices index and show, we want to give the candidate contact information for the provider of the course.

The information appears on these pages, with these statuses:

**Application review page** statuses:
    `awaiting_provider_decision`, `inactive`, `interviewing` or `offer`,

**Your offer page** statuses:
    `offer_deferred`, `pending_conditions` or `recruited`,

## Guidance to review

New pattern introduced to the system specs:
  create a local helper file and `require_relative`.
  The helpers will be specific to the directory so no need to mix them in with RSpec globally or use shared examples

### Screenshots

_The spacing of the elements shown here are due to change in a later PR aimed at standardising the CSS and markup applied to these elements. These screenhots show that the relevant information is displayed with the correct content on the right pages._

|Application Page|Offer Page|
|---|---|
|Screenshot from 2024-02-07 14-55-33|Screenshot from 2024-02-07 14-24-16|
## Link to Trello card

[Trello Ticket](https://trello.com/c/BPUIGcx7/1172-apply-contact-training-provider-section-part-32)

### CODE REVIEW

**Comment by inulty-dfe** [comment]
[Apply: Contact training provider section (part 3.2)](https://trello.com/c/BPUIGcx7/1172-apply-contact-training-provider-section-part-32)

**Review by lsalivonaite-transform — CHANGES_REQUESTED** [review]
I don't think we need the 'contact training provider' section on the 'conditions_not_met' status. 

Also could we rearrange the sections so that the contact training provider section would be at the bottom and not in between 'what happens next' and 'withdraw application'? So it should be:
1. What happens next
2. Withdraw your application
3. Contact training provider

**Comment by inulty-dfe** [comment]
> I don't think we need the 'contact training provider' section on the 'conditions_not_met' status.
> 
> Also could we rearrange the sections so that the contact training provider section would be at the bottom and not in between 'what happens next' and 'withdraw application'? So it should be:
> 
> 1. What happens next
> 2. Withdraw your application
> 3. Contact training provider

I've updated the order now, and the component does not appear for `conditions_not_met`.

**Review by lsalivonaite-transform — APPROVED** [review]

File: app/controllers/candidate_interface/offer_dashboard_controller.rb [hunk 1]
```diff
@@ -10,6 +10,7 @@ def show
       @application_choice = choices.pending_conditions.first || choices.recruited.first || choices.offer_deferred.first
       @provider_name = @application_choice.current_provider.name
       @course_name_and_code = @application_choice.current_course.name_and_code
+      @provider = @application_choice.current_provider
```
Comment (vassyz) [inline app/controllers/candidate_interface/offer_dashboard_controller.rb:13]: Should we then remove line 11 and replace `@provider_name` with `@provider.name`?

Comment (vassyz) [reply app/controllers/candidate_interface/offer_dashboard_controller.rb:13]: Or similar to what I suggested in my previous comment, using `@application_choice` in the component.

File: spec/components/candidate_interface/continuous_applications/application_review_component_spec.rb [hunk 2]
```diff
@@ -120,6 +120,14 @@
     it 'does not show withdraw CTA' do
       expect(result.text).not_to include('withdraw this application')
     end
+
+    describe 'does not show provider contact information' do
```
[the remark this answers is not in the document]
Comment (vassyz) [reply spec/components/candidate_interface/continuous_applications/application_review_component_spec.rb:124]: Can we also have `it 'shows` or something that reads better?

File: app/components/candidate_interface/continuous_applications/application_review_component.rb [hunk 3]
```diff
@@ -160,6 +160,10 @@ def show_withdraw?
       def can_add_more_choices?
         application_choice.application_form.can_add_more_choices?
       end
+
+      def provider
```
[the remark this answers is not in the document]
Comment (inulty-dfe) [reply app/components/candidate_interface/continuous_applications/application_review_component.rb:164]: I'll make a method `def status`

[the remark this answers is not in the document]
Comment (inulty-dfe) [reply app/components/candidate_interface/continuous_applications/application_review_component.rb:164]: Arguably, status shouldn't be passed into the component..

File: app/components/candidate_interface/provider_contact_information_component.html.erb [hunk 4]
```diff
@@ -0,0 +1,11 @@
+<div class="govuk-!-margin-bottom-6">
+  <h2 class="govuk-heading-m govuk-!-margin-bottom-1">Contact training provider</h2>
+  <p class="govuk-body">Contact <%= provider.name %> if you have any questions about your application.</p>
+  <% if provider.phone_number && provider.email_address %>
```
[the remark this answers is not in the document]
Comment (inulty-dfe) [reply app/components/candidate_interface/provider_contact_information_component.html.erb]: Moved to a method and tested

File: spec/components/candidate_interface/provider_contact_information_component_spec.rb [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+  end
+
+  context 'when provider has only phone' do
+    let(:provider) do
+      create(:provider, email_address: nil)
+    end
+
+    it 'show provider phone number' do
+      expect(result.text).to include('Contact training provider',
+                                     "Call on #{provider.phone_number}")
```
Comment (tomas-stefano) [inline spec/components/candidate_interface/provider_contact_information_component_spec.rb:23]: Did you consider adding a not to include "Email at" here? and not to include "Contact" below? Do you think it is worth?

**Review by tomas-stefano — APPROVED** [review]
One small question/consideration but code looks good 🚀

File: spec/components/candidate_interface/provider_contact_information_component_spec.rb (hunk 5 above)
Comment (inulty-dfe) [reply spec/components/candidate_interface/provider_contact_information_component_spec.rb:23]: I don't think it's worth it but I'm open to being pushed

File: app/components/candidate_interface/continuous_applications/application_review_component.rb [hunk 6]
```diff
@@ -157,9 +157,24 @@ def show_withdraw?
         ApplicationStateChange.new(@application_choice).can_withdraw?
       end
 
+      def show_provider_contact_details?
+        application_states = %w[
+          awaiting_provider_decision
+          inactive
+          interviewing
+          offer
+        ].freeze
```
Comment (vassyz) [inline app/components/candidate_interface/continuous_applications/application_review_component.rb:166]: Should this have `.freeze`?

File: app/controllers/candidate_interface/offer_dashboard_controller.rb [hunk 7]
```diff
@@ -8,10 +8,18 @@ def show
       @application_form = current_application
       choices = current_application.application_choices.includes(:offer, course_option: [course: :provider])
       @application_choice = choices.pending_conditions.first || choices.recruited.first || choices.offer_deferred.first
-      @provider_name = @application_choice.current_provider.name
+      @provider = @application_choice.current_provider
       @course_name_and_code = @application_choice.current_course.name_and_code
     end
 
+    helper_method def show_provider_contact_information?
```
Comment (vassyz) [inline app/controllers/candidate_interface/offer_dashboard_controller.rb:15]: Other method is called `show_provider_contact_details?`. Not that it matters.

File: spec/components/candidate_interface/provider_contact_information_component_spec.rb (hunk 5 above)
Comment (vassyz) [reply spec/components/candidate_interface/provider_contact_information_component_spec.rb:23]: Not a big issue, but I agree with @tomas-stefano. If you don't want to add not include "Email at" and "Call on" I'd remove these two specs as they are already being covered by the first spec. They are only testing if you are not somehow hiding the phone if the email is `nil`, which is an unlikely scenario.

File: spec/system/candidate_interface/post_offer/offer_deferred_spec.rb [hunk 8]
```diff
@@ -0,0 +1,35 @@
...
+RSpec.feature 'Post-offer dashboard' do
+  include CandidateHelper
+  include PostOfferHelper
+
+  scenario 'Candidate offer is deferred' do
+    given_i_am_signed_in
+    and_i_have_an_accepted_offer_deferred
+
+    when_i_visit_the_application_dashboard
+    then_i_see_that_i_have_defered_my_offer
```
Comment (vassyz) [inline spec/system/candidate_interface/post_offer/offer_deferred_spec.rb:13]: `deferred`

File: spec/components/candidate_interface/provider_contact_information_component_spec.rb (hunk 5 above)
Comment (inulty-dfe) [reply spec/components/candidate_interface/provider_contact_information_component_spec.rb:23]: I've added it

File: app/components/candidate_interface/continuous_applications/application_review_component.rb (hunk 6 above)
Comment (inulty-dfe) [reply app/components/candidate_interface/continuous_applications/application_review_component.rb]: Removed it

**Pull request merged**